relay-server/src/services/buffer/mod.rs

```diff
@@ -81,6 +86,92 @@ impl FromMessage<Self> for EnvelopeBuffer {
     }
 }
 
+/// Abstraction that wraps a list of [`ObservableEnvelopeBuffer`]s to which [`Envelope`] are routed
+/// based on their [`ProjectKeyPair`].
+#[derive(Debug, Clone)]
+pub struct PartitionedEnvelopeBuffer {
+    buffers: Arc<Vec<ObservableEnvelopeBuffer>>,
+}
+
+impl PartitionedEnvelopeBuffer {
+    /// Creates a [`PartitionedEnvelopeBuffer`] with no partitions.
+    #[cfg(test)]
+    pub fn empty() -> Self {
+        Self {
+            buffers: Arc::new(Vec::new()),
+        }
+    }
+
+    /// Creates a new [`PartitionedEnvelopeBuffer`] by instantiating inside all the necessary
+    /// [`ObservableEnvelopeBuffer`]s.
+    #[allow(clippy::too_many_arguments)]
+    pub fn create(
+        partitions: NonZeroU8,
+        config: Arc<Config>,
+        memory_stat: MemoryStat,
+        global_config_rx: watch::Receiver<global_config::Status>,
+        envelopes_tx: mpsc::Sender<legacy::DequeuedEnvelope>,
+        project_cache_handle: ProjectCacheHandle,
+        outcome_aggregator: Addr<TrackOutcome>,
+        test_store: Addr<TestStore>,
+        runner: &mut ServiceRunner,
+    ) -> Self {
+        let mut envelope_buffers = Vec::with_capacity(partitions.get() as usize);
+        for partition_id in 0..partitions.get() {
+            let envelope_buffer = EnvelopeBufferService::new(
+                partition_id,
+                config.clone(),
+                memory_stat.clone(),
+                global_config_rx.clone(),
+                Services {
+                    envelopes_tx: envelopes_tx.clone(),
+                    project_cache_handle: project_cache_handle.clone(),
+                    outcome_aggregator: outcome_aggregator.clone(),
+                    test_store: test_store.clone(),
+                },
+            )
+            .map(|b| b.start_in(runner));
+
+            if let Some(envelope_buffer) = envelope_buffer {
+                envelope_buffers.push(envelope_buffer);
+            }
+        }
+
+        Self {
+            buffers: Arc::new(envelope_buffers),
+        }
+    }
```

Comment:
You can do this with `collect`, though it's probably a matter of taste which is better.
```suggestion
        let envelope_buffers = (0..partitions.get())
            .filter_map(|partition_id| {
                EnvelopeBufferService::new(
                    partition_id,
                    config.clone(),
                    memory_stat.clone(),
                    global_config_rx.clone(),
                    Services {
                        envelopes_tx: envelopes_tx.clone(),
                        project_cache_handle: project_cache_handle.clone(),
                        outcome_aggregator: outcome_aggregator.clone(),
                        test_store: test_store.clone(),
                    },
                )
            })
            .map(|b| b.start_in(runner))
            .collect();
```